Support discovery REST API in sidecar agent

craigyam opened the pull request:
Instead of the sidecar pushing new updates to the proxy (whether its
nginx or envoy), the sidecar could also listen on a localhost endpoint
where the proxy could poll periodically to obtain upstream instances
(the registry data). This model would decouple the proxy implementation
from the sidecar agent and allow us to switch between different proxies
as well.

At minimum, the sidecar needs to support Envoy's API endpoint for
service discovery.

craigyam committed:
Support discovery REST API in sidecar agent

Instead of the sidecar pushing new updates to the proxy (whether its
nginx or envoy), the sidecar could also listen on a localhost endpoint
where the proxy could poll periodically to obtain upstream instances
(the registry data). This model would decouple the proxy implementation
from the sidecar agent and allow us to switch between different proxies
as well.

At minimum, the sidecar needs to support Envoy's API endpoint for
service discovery.

craigyam committed:
Add a short sleep after stopping the server.

Need to sleep a bit in the teardown routine to ensure the
server is stopped before the next test.

craigyam committed:
Run server in go routine.

The serve() blocks so needs to be run in go routine.  Also,
move the start of the server inside the proxy check since that's
when it will be needed.

ijsnellf commented on sidecar/sidecar.go:
```diff
@@ -101,6 +101,25 @@ func Run(conf config.Config) error {
 			return err
 		}
+
+		if conf.DiscoveryPort == 0 {
```
Instead of setting the default here, can you remove this segment of code and instead put the default inside of `sidecar/config/default.go`? That way we support config files, env vars, and args.

ijsnellf commented on sidecar/config/config.go:
```diff
@@ -256,6 +258,7 @@ func (c *Config) loadFromContext(context *cli.Context) error {
 	loadFromContextIfSet(&c.Dnsconfig.Domain, dnsConfigDomainFlag)
 	loadFromContextIfSet(&c.LogLevel, logLevelFlag)
 	loadFromContextIfSet(&c.Debug, debugFlag)
+	loadFromContextIfSet(&c.DiscoveryPort, discoveryPortFlag)
```
Could you add a validator for the discovery port to the `Validate` function? Something like `IsInRange("Service Discovery Port", c.DiscoveryPort, 1, 65535)`.

craigyam replied:
Done

ijsnellf commented on sidecar/register/discovery/discovery.go:
```diff
@@ -0,0 +1,174 @@
+type Host struct {
+	IPAddr string            `json:"ip_address"`
+	Port   uint16            `json:"port"`
+	Tags   map[string]string `json:"tags"`
```
Alternately, we know that Envoy only supports `az`, `canary` and `load_balance_weight` fields, so we can explicitly declare them here. Not necessary for this PR, though.

Reference:
https://lyft.github.io/envoy/docs/configuration/cluster_manager/sds_api.html#config-cluster-manager-sds-api

ijsnellf commented on sidecar/register/discovery/discovery.go:
```diff
@@ -0,0 +1,174 @@
+	return
+}
+
+func translate(instances []*api.ServiceInstance) []Host {
```
This logic will be expanded. This looks good for a simple 1st pass.

ijsnellf commented on sidecar/register/server.go:
```diff
@@ -0,0 +1,136 @@
+	}
+
+	s := &server{
+		config: &*conf,
```
What does this do?

craigyam replied:
Hm, not sure.  Took this from registry.  Seems to work fine without the &* so I took it out.

ijsnellf commented on sidecar/register/server.go:
```diff
@@ -0,0 +1,136 @@
+func (s *server) Stop() {
+	s.logger.Info("Stopping rest server")
+
+	if err := s.listener.Close(); err != nil {
```
Should we ensure `s.listener != nil`?

craigyam replied:
Done

ijsnellf commented on sidecar/register/server.go:
```diff
@@ -0,0 +1,136 @@
+		&rest.ContentTypeCheckerMiddleware{},
+	)
+
+	log.SetOutput(s.logger.Logger.Out)
```
Doesn't this override all log output except for `s.logger.Logger.Out` output? The `registry/util/logging` package should already be directing output correctly, I thought.

craigyam replied:
Done

ijsnellf reviewed (changes requested):
Looks pretty good. Few comments / recommendations.

craigyam committed:
Process code review comments.

- Add default port to default.go
- Add validator for discovery port
- Update config unit tests for discovery port default
- Clean up a few things in server.go

craigyam committed:
Merge branch 'master' into sidecar_discovery_api

ijsnellf reviewed (approved):
LGTM, but let me create an envoy branch to push this to instead of master.
